[CIR] Upstream CmpOp

mmha opened the pull request:
This patch adds support for comparison operators with ClangIR, both integral and floating point.

mmha commented:
cc @andykaylor @erichkeane @dkolsen-pgi

mmha committed:
clang-format

andykaylor reviewed (commented):
This looks really good. I have just a few minor comments and a request for more tests.

bcardosolopes reviewed (approved):
I don't have anything to add on top of Andy's review. LGTM once those comments are addressed.

xlauko commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -826,6 +826,50 @@ def ForOp : CIR_Op<"for", [LoopOpInterface, NoRegionArguments]> {
+  let assemblyFormat = [{
+    `(` $kind `,` $lhs `,` $rhs  `)` `:` type($lhs) `,` type($result) attr-dict
+  }];
+
+  // Already covered by the traits
+  let hasVerifier = 0;
```
No need, this is default value.

```suggestion
```

xlauko commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -710,6 +710,89 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+
+    return emitScalarConversion(result, cgf.getContext().BoolTy, e->getType(),
+                                e->getExprLoc());
+  }
```
There is a bit of a mess how locations are emitted here:

`cgf.getLoc(e->getExprLoc())` vs. `cgf.getLoc(boInfo.loc)` vs. `cgf.getLoc(e->getSourceRange())`

tbf I am not sure what is correct way, or whether clangir even has some rule of thumb how AST locations are translated to mlir locations. Just the inconsistency caught my eye. @bcardosolopes @andykaylor ?

mmha replied:
FWIW classic codegen uses `getExprLoc` and doesnt emit a range.

andykaylor replied:
I agree that we need to look at context to decide which source location/range makes sense. I've seen problems where we were getting source locations from operands rather than from the expression where the operand is used, which was wrong in that case.

mmha commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -710,6 +710,89 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+        assert(!cir::MissingFeatures::fixedPointType());
+        cgf.cgm.errorNYI(boInfo.loc, "fixed point comparisons");
+        result = builder.getBool(false, cgf.getLoc(boInfo.loc));
+      } else if (lhsTy->hasSignedIntegerRepresentation()) {
```
I think it's better to remove that branch altogether and let the `else` deal with it.

mmha commented on clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp:
```diff
@@ -514,9 +515,17 @@ mlir::LogicalResult CIRToLLVMCastOpLowering::matchAndRewrite(
     break;
-  case cir::CastKind::ptr_to_bool:
-    assert(!cir::MissingFeatures::opCmp());
+  case cir::CastKind::ptr_to_bool: {
```
I wanted to ask you the same question so I guess the answer is no :D

andykaylor replied:
As far as I know, there is no additional benefit to going through the cir ops.

@bcardosolopes Can you comment on this?

bcardosolopes replied:
Sometimes we'd go for the CIR ops because they already have a canonical way to lower and if something changes there, we get the benefits elsewhere (there are also cases where it might be confusing to look at the adaptor versus the CIR operation in other to take a decision, in which case fallbacking to known CIR lowering also helps). This change doesn't seem related to those cases and LGTM.

mmha committed:
Apply suggestions from code review

Co-authored-by: Henrich Lauko <xlauko@mail.muni.cz>
Co-authored-by: Andy Kaylor <akaylor@nvidia.com>

mmha committed:
Address review feedback
- Expand testing

mmha commented on clang/test/CIR/CodeGen/cmp.cpp:
```diff
@@ -0,0 +1,233 @@
+// RUN: %clang_cc1 -std=c++17 -triple x86_64-unknown-linux-gnu -fclangir -emit-cir -DCIR_ONLY %s -o %t.cir
+// RUN: FileCheck --input-file=%t.cir %s -check-prefix=CIR
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -Wno-unused-value -fclangir -emit-llvm %s -o %t-cir.ll
+// RUN: FileCheck --input-file=%t-cir.ll %s -check-prefix=LLVM
```
Done. I also added tests for int/ptr to bool in the casting tests.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -710,6 +710,85 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+        cgf.cgm.errorNYI(loc, "fixed point comparisons");
+        result = builder.getBool(false, loc);
+      } else {
+        // Unsigned integers and pointers.
```
```suggestion
        // integers and pointers
```

erichkeane reviewed (commented):
No comments on my run through, so Andy-et-al + Bruno is enough for me.

bcardosolopes reviewed (approved):
LGTM after you add a roundtrip test for `cir::CmpOp` (in a .cir file).

mmha committed:
[CIR] Upstream CmpOp

This patch adds support for comparison operators with ClangIR, both integral and floating point.

mmha committed:
Review feedback

- Fix comment
- Add roundtrip test

andykaylor reviewed (approved):
lgtm
